### ISSUE TITLE
Why does the model project relatively static pixels to infinity?

### ISSUE THREAD

**Issue by aj96** [nianticlabs/monodepth2#42]
I thought I'd open up a discussion.

I recently tried training monodepth2 on cityscapes and discovered that even with static mask, it was failing to mask out some vehicles moving with the ego car and was projecting them to infinity. In order to resolve this issue, I want to understand exactly why projecting these pixels to infinity minimizes the loss. This is what I don't entirely understand. The best understanding I have is that in order to make a point in 3D not move across frames, the easiest way to do this is to set its z-component to infinity, so that the euclidean distance between the two points zero. The z-component would be the depth from the depth map. But somehow, I don't believe this is right. Does anyone have a clear mathematical formulation that shows, given the re-projection equation, minimizing the loss comes from setting z to infinity? I believe that if I know the exact mathematical component that is causing this issue, it will be clearer how to resolve the issue.

The most obvious thing to try would be to come up with a different loss function to pass the static pixels through to prevent them from being projected to infinity. But then you'd need some sort of idea of what the actual depth should be, and considering this is unsupervised approach, I don't see how you can do that. I know struct2depth added a regularizing term to approximate what the depth of an object should be by learning the objects height and using similar triangles. But I tried struct2depth as well and it was still projecting objects to infinity, although this might have been due to the fact that MASK-RCNN was missing too many objects. Just throwing ideas out there. Thanks.

**Comment by larsme** [comment]
From what I understand, infinite depth only makes sense if there is no rotation.
The matrix product becomes an identity matrix,
the translation is negligible because its so much smaller
1/z gets applied because of homogeneaus coordinates
and you get p'=p.

If there is a (small) rotation, i assume infinity might still be a possible solution for low texture regions.

**Comment by aj96** [comment]
But for translation to be small and for rotation matrix to be identity matrix that would mean that all the pixels didn't move. That doesn't explain the case for one of relatively static pixels like a car that moves with the ego-car.

**Comment by larsme** [comment]
Its not that translation is small, just smaller than the "competing" term scaled by near infinite depth

**Comment by daniyar-niantic** [comment]
@aj96 I think you are more-or-less on the right track with the reprojection equation. You just need to take it a step further and look at the projection of 3D point into the 2D view and look at the pixel coordinates of relatively static pixel. 

Another way to look at this is with an extreme situation, where ego car is following car that is moving at the same speed as ego car. Let's assume there is no rotation for now, i.e. rotation matrix is identity. If you consider a point on a car that is moving with the ego car, then that point will have exactly the same pixel coordinates in both view. So, if you were to rectify these two views using ground truth poses, then the relatively static pixel will have 0 disparity. Finally, 0 disparity is equivalent to infinite depth.
